src/Functions/FunctionsConversion.h

```diff
@@ -1126,75 +1128,85 @@ class FunctionConvertFromString : public IFunction
-                this->getName().find("OrNull") != std::string::npos)
-                throw Exception("Illegal type " + arguments[0].type->getName() + " of first argument of function " + getName() +
-                        ". Conversion functions with postfix 'OrZero' or 'OrNull'  should take String argument",
-                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
-            else
-                throw Exception("Illegal type " + arguments[0].type->getName() + " of first argument of function " + getName(),
-                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
-        }
+            validateFunctionArgumentTypes(*this, arguments,
+                FunctionArgumentDescriptors{{"string", isStringOrFixedString, nullptr, "String or FixedString"}},
+                // optional
+                FunctionArgumentDescriptors{
+                    {"precision", isUInt8, isColumnConst, "const UInt8"},
+                    {"timezone", isStringOrFixedString, isColumnConst, "const String or FixedString"},
+                });
 
-        if (arguments.size() == 2)
+            UInt64 scale = DataTypeDateTime64::default_scale;
+            if (arguments.size() > 1)
+                scale = extractToDecimalScale(arguments[1]);
+            const auto timezone = extractTimeZoneNameFromFunctionArguments(arguments, 2, 0);
+            res = std::make_shared<DataTypeDateTime64>(scale, timezone);
+        }
+        else
         {
-            if constexpr (std::is_same_v<ToDataType, DataTypeDateTime>)
-            {
-                if (!isString(arguments[1].type))
-                    throw Exception("Illegal type " + arguments[1].type->getName() + " of 2nd argument of function " + getName(),
-                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
-            }
-            else if constexpr (to_decimal)
+            if ((arguments.size() != 1 && arguments.size() != 2) || (to_decimal && arguments.size() != 2))
+                throw Exception("Number of arguments for function " + getName() + " doesn't match: passed " + toString(arguments.size()) +
+                    ", should be 1 or 2. Second argument only make sense for DateTime (time zone, optional) and Decimal (scale).",
+                    ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+            if (!isStringOrFixedString(arguments[0].type))
             {
-                if (!isInteger(arguments[1].type))
-                    throw Exception("Illegal type " + arguments[1].type->getName() + " of 2nd argument of function " + getName(),
-                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
-                if (!arguments[1].column)
-                    throw Exception("Second argument for function " + getName() + " must be constant", ErrorCodes::ILLEGAL_COLUMN);
+                if (this->getName().find("OrZero") != std::string::npos ||
+                    this->getName().find("OrNull") != std::string::npos)
```

Comment:
`endsWith()` seems to be more suitable here.

Reply:
Yes, but I've just tossed that under an else-branch of `if constexpr`, so that piece is mostly as it was before.

IMO that exact piece (and the whole approach of how compile-time branching based on function return type is done) should be re-thought and re-designed, this piece is not the worst, but it is becoming a mess.